nuclia/sdk/upload.py

```diff
@@ -30,144 +59,211 @@ def _auth(self) -> NucliaAuth:
         with requests.get(origin, stream=True) as r:
             filename = origin.split("/")[-1]
             size_str = r.headers.get("Content-Length")
             if size_str is None:
                 size_str = "-1"
             size = int(size_str)
-            upload_url = ndb.start_tus_upload(
-                rid=rid,
-                field=field,
-                size=size,
-                filename=filename,
-                content_type=r.headers.get("Content-Type", "application/octet-stream"),
+            mimetype = r.headers.get("Content-Type", "application/octet-stream")
+            rid, is_new_resource = self._get_or_create_resource(
+                rid=rid, icon=mimetype, **kwargs
             )
-            offset = 0
-            for _ in tqdm(range((size // 524288) + 1)):
-                chunk = r.raw.read(524288)
-                offset = ndb.patch_tus_upload(upload_url, chunk, offset)
+            try:
+                upload_url = ndb.start_tus_upload(
+                    rid=rid,
+                    field=field,
+                    size=size,
+                    filename=filename,
+                    content_type=mimetype,
+                )
+                offset = 0
+                for _ in tqdm(range((size // 524288) + 1)):
+                    chunk = r.raw.read(524288)
+                    offset = ndb.patch_tus_upload(upload_url, chunk, offset)
+            except Exception as e:
+                print(e)
+                if is_new_resource:
+                    ndb.ndb.delete_resource(kbid=ndb.kbid, rid=rid)
+                sys.exit(1)
+
+    def _get_or_create_resource(*args, **kwargs) -> Tuple[str, bool]:
+        rid = kwargs.get("rid")
+        if rid:
+            return rid
+        ndb = kwargs["ndb"]
+        slug = kwargs.get("slug")
+        need_to_create_resource = slug is None
+        if slug:
+            try:
+                resource = ndb.ndb.get_resource_by_slug(kbid=ndb.kbid, slug=slug)
+                rid = resource.id
+                logger.warn(f"Using existing resource: {rid}")
+                need_to_create_resource = False
+            except exceptions.NotFoundError:
+                need_to_create_resource = True
+        else:
+            slug = uuid4().hex
+
+        if need_to_create_resource:
+            kw = {
+                "kbid": ndb.kbid,
+                "slug": slug,
+            }
+            for param in RESOURCE_ATTRIBUTES:
+                if kwargs.get(param):
+                    kw[param] = kwargs.get(param)
+            resource = ndb.ndb.create_resource(**kw)
+            rid = resource.uuid
+            logger.warn(f"New resource created: {rid}")
+
+        return (rid, need_to_create_resource)
+
+    def _update_resource(self, rid: str, **kwargs):
+        ndb = kwargs["ndb"]
+        kw = {
+            "kbid": ndb.kbid,
+            "rid": rid,
+        }
+        for param in RESOURCE_ATTRIBUTES:
+            if kwargs.get(param):
```

Comment:
I would change it to `if param in kwargs:` so that you are also able to set values to None or "" if needed